sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,263 @@
+        /// <param name="connection">The <see cref="EventHubConnection"/> to use as the basis for delegation of client-type operations.</param>
+        /// <param name="retryPolicy">The <see cref="EventHubsRetryPolicy"/> to use as the basis for retrieving the information.</param>
+        ///
+        /// <returns>A <see cref="PooledProducer"/> corresponding to the partitionId passed in as parameter.</returns>
+        ///
+        /// <remarks>
+        ///   There is a slight probability that the returned producer may be closed at any time
+        ///   after it is returned and the caller may want to handle that scenario.
+        /// </remarks>
+        ///
+        public PooledProducer GetPartitionProducer(string partitionId,
+                                                   EventHubConnection connection,
+                                                   EventHubsRetryPolicy retryPolicy)
+        {
+            Argument.AssertNotNull(partitionId, nameof(partitionId));
+            Argument.AssertNotNull(connection, nameof(connection));
+            Argument.AssertNotNull(retryPolicy, nameof(retryPolicy));
+
+            var identifier = Guid.NewGuid().ToString();
+
+            var item = Pool.GetOrAdd(partitionId, id => new PoolItem(connection.CreateTransportProducer(id, retryPolicy)));
+
+            // A race condition at this point may end with an CloseAsync called on
+            // the returned PoolItem if it had expired. The probability is very low and
+            // possible exceptions should be handled by the invoking methods.
+
+            if (!item.ActiveInstances.TryAdd(identifier, 0) || item.Producer.IsClosed)
+            {
+                identifier = Guid.NewGuid().ToString();
+                item = Pool.GetOrAdd(partitionId, id => new PoolItem(connection.CreateTransportProducer(id, retryPolicy)));
+                item.ActiveInstances.TryAdd(identifier, 0);
+            }
+
+            item.RemoveAfter = DateTimeOffset.UtcNow.Add(PoolItem.DefaultRemoveAfterDuration);
+
+            return new PooledProducer(item.Producer, cleanUp: producer =>
+            {
+                if (Pool.TryGetValue(partitionId, out PoolItem pooledItem))
+                {
+                    pooledItem.RemoveAfter = DateTimeOffset.UtcNow.Add(PoolItem.DefaultRemoveAfterDuration);
+                }
+
+                if (!item.ActiveInstances.TryRemove(identifier, out _))
+                {
+                    // The PoolItem will not be closed deterministically
+                    // and the ExpirationTimer callback will eventually remove it from the
+                    // Pool leaving to the Garbage Collector the responsability of closing
+                    // the TransportProducer and the AMQP link.
+                }
+
+                if (!Pool.TryGetValue(partitionId, out _) && !item.ActiveInstances.Any())
```

Comment:
Do we want `TryGetValue` here or `TryRemove`?

Reply:
Quoting you:

>  I'm probably being overly defensive here, but the second
>  TryGet runs after the extension would have been seen, so it 
>  is intended to be sure that the item wasn't removed in the meantime.

It personally looks ok to me, but I may be missing some pieces.

Reply:
I the meantime, I added a comment that would work as reminder.